pydantic/main.py

```diff
@@ -367,7 +367,8 @@ def __setattr__(self, name, value):
         elif self.__config__.validate_assignment:
             known_field = self.__fields__.get(name, None)
             if known_field:
-                value, error_ = known_field.validate(value, self.dict(exclude={name}), loc=name, cls=self.__class__)
+                values = dict(self._iter(exclude={name}))
+                value, error_ = known_field.validate(value, values, loc=name, cls=self.__class__)
                 if error_:
                     raise ValidationError([error_], self.__class__)
```

Comment:
Nice catch ! I wonder if it wouldn't be faster and more explicit to just write
```python
original_value = self.__dict__.pop(name)
value, error_ = known_field.validate(value, self.__dict__, loc=name, cls=self.__class__)
if error_:
  self.__dict__[name] = original_value
  raise ValidationError([error_], self.__class__)
...
```

Reply:
Honestly I didn't went into what's going on in `_iter`, I just saw the `as_dict` param and thought changing it to False would fix the problem.
Your suggestion is better, I agree.